TST: stats: ensure tests are thread-safe

andfoy opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
See https://github.com/scipy/scipy/issues/20669

#### What does this implement/fix?
<!--Please explain your changes.-->
This is a continuation of the work done in https://github.com/scipy/scipy/pull/21496, this time, each module gets its own PR detailing changes that ensure that the tests pass under a concurrent scenario. 

#### Additional information
<!--Any additional information you think is important.-->
See the description of https://github.com/scipy/scipy/pull/21496 for more detailed information.

cc @rgommers

andfoy committed:
FIX: Make scipy.stats continuous/discrete distributions tests to work under concurrent loads

andfoy committed:
FIX: fix remaining tests

rgommers reviewed (commented):
Thanks @andfoy! Looks good overall.

I did an initial browse through the diff, and the vast majority of changes are to the tests - the few changes to the `stats` functionality look pretty straightforward to understand. I plan to review and test in detail and get this in, then backport only the functional changes needed for thread-safety to 1.15.x and not the test changes (@tylerjereddy please comment if you don't agree).

@mdhaber do you want to review (parts of) this PR, or are you happy for me to do it?

rgommers commented:
Three of the CI failures need resolving, but they're all minor.

tylerjereddy commented:
>  I plan to review and test in detail and get this in, then backport only the functional changes needed for thread-safety to 1.15.x and not the test changes (@tylerjereddy please comment if you don't agree).

Seems "ok" I guess if you're objective is to avoid crazy merge conflicts (and `stats` already suffers from that a fair bit because of all the overhauling that happened).

I suppose if we had a ton of bug reports in `1.15.0` related to free-threading then backporting the test changes might make sense, though I believe we'd labeled the support as prelim in any case.

mdhaber commented:
> do you want to review (parts of) this PR, or are you happy for me to do it?

Both. Feel free to merge on your schedule, but I'll look at some point. When we started discussing caching in the context of `different`/`tanhsinh`, I realized we'd need to do something about MWU (and maybe other exact tests). Hopefully this can use a cache safely or show that it's no longer important for performance? (It might not be since we included the much faster MWU algorithm.)

mdhaber commented on scipy/stats/tests/test_axis_nan_policy.py:
```diff
@@ -598,7 +602,7 @@ def test_keepdims(hypotest, args, kwds, n_samples, n_outputs, paired, unpacker,
         def unpacker(res):
             return res
-    rng = np.random.default_rng(0)
+    rng = np.random.RandomState(0)
```
Why change from the more modern `default_rng` to `RandomState`?

rgommers replied:
I changed it back, let's see if CI is happy. I assume this was done in the context of a test failure. A lot of tests rely on the exact random numbers used, and this looks like an alternative way of fixing such failures (changing the seed or bumping the tolerance are the other options).

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2601,16 +2604,13 @@ def test_sf(self):
         expected = (scale/x)**b   # 2.25e-18
         assert_allclose(p, expected)
 
-    @pytest.fixture(scope='function')
```
I had not used a fixture like this before but just started to. We can't do this?

rgommers replied:
No, this is not thread-safe. It would create `rng` only once per test, and then draw samples from it multiple times when run under `pytest-run-parallel`. So the random numbers change in different threads, making the test fail when tolerances are set tightly.

It also looks like a micro-optimization that makes the code harder to understand, so I would avoid such constructs.

rgommers replied:
If we really need a fixture like this for something, it can still be used - but then the test itself should be marked with the `thread_unsafe` marker.

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -2739,6 +2740,10 @@ def ansari(x, y, alternative='two-sided'):
         raise ValueError("'alternative' must be 'two-sided',"
                          " 'greater', or 'less'.")
+
+    if not hasattr(_abw_state, 'a'):
+        _abw_state.a = _ABW()
```
Could you add a comment in the code about how this works and why the name of the attribute is `a` (or `s` in some tests)? Same with the MWU state. Thanks!

rgommers replied:
See https://docs.python.org/3/library/threading.html#thread-local-data. This pattern is the same everwhere - avoid mutable global state. Note that the `.a` name is irrelevant. I think @andfoy has chosen one-letter names to keep line lengths short; it could also have been named `.val` or something like that.

The more detailed explanation here: `_ABW` is a class whose methods mutate attributes of the instance. If you have an `_ABW` instance that is global to the module, then there is nothing preventing (in the absence of the GIL) parallel calls to `_ABW` methods from multiple threads. That clearly is going to go wrong (e.g., thread 1 is in the middle of something, and thread 2 overwrites `self.freqs`). So the solution is to make the stateful thing thread-local.

rgommers replied:
I added these comments, and included them in the PR split off from this one (gh-22175).

mdhaber commented on scipy/stats/tests/test_resampling.py:
```diff
@@ -332,14 +332,15 @@ def test_bootstrap_against_itself_1samp(method, expected):
                     method=method,
-                    axis=-1)
+                    axis=-1,
+                    random_state=rng)
```
Due to [SPEC 7](https://scientific-python.org/specs/spec-0007/), Please use `rng` instead of `random_state` where possible. (That's pretty much everwhere that accepts `random_state` or `seed` right now *except* the old distribution infrastructure, like the call to `dist.rvs` above.

On a related note, please use of `default_rng` instead of `RandomState` to generate instances that get passed into functions when there is a choice. Internally, these functions will (in the future) use `default_rng` to normalize the `rng` argument, and until NumPy 2.2, `default_rng` will not accept an instance of `RandomState` as an argument.

rgommers replied:
Renamed from `random_state` to `rng`, resolving.

mdhaber reviewed (commented):
Thanks for this!

tylerjereddy reviewed (commented):
I checked locally on ARM Mac laptop with test incantation `python dev.py test -j 8 -t scipy/stats -- --parallel-threads=10` and found that the latest `main` branch had several failures and seemed to end up hanging or stuck for a long time on some tests.

On this feature branch, I didn't see any signs of hanging/slow performance, but I still observed some test failures (157 in total).

Most, perhaps all, of the failures are related to failure to issue a warning in a `stats` test and IIRC warnings handling has thread safety issues?

Regardless, here is a sample of one the fails I see locally:

```
_______________________________________________________________________ TestVariation.test_combined_edge_cases[omit-numpy] ________________________________________________________________________
[gw5] darwin -- Python 3.13.0 /Users/treddy/python_venvs/py_313_free_threaded/bin/python
scipy/stats/tests/test_variation.py:173: in test_combined_edge_cases
    with pytest.warns(SmallSampleWarning, match=too_small_nd_omit):
E   Failed: DID NOT WARN. No warnings of type (<class 'scipy.stats._axis_nan_policy.SmallSampleWarning'>,) were emitted.
E    Emitted warnings: [].
        nan_policy = 'omit'
        self       = <scipy.stats.tests.test_variation.TestVariation object at 0x2eee809b170>
        x          = array([[ 0., 10., nan,  1.],
       [ 0., -5., nan,  2.],
       [ 0., -5., nan,  3.]])
        xp         = <module 'numpy' from '/Users/treddy/python_venvs/py_313_free_threaded/lib/python3.13t/site-packages/numpy/__init__.py'>
        y          = array([       nan,        inf,        nan, 0.40824829])
```

Edit: `export PYTHON_GIL=0 ` always set for above checks

rgommers commented:
> Most, perhaps all, of the failures are related to failure to issue a warning in a stats test and IIRC warnings handling has thread safety issues?

Yes that is correct. For the time being we've marked threads that check warnings with `thread_unsafe`. Hopefully in Python 3.14 we will finally get a thread-safe `warnings` module (it's being investigated), so fingers crossed.

Since the PR for `1.15.0rc2` backports is already up, I'm going to split off the functional changes that need backporting first. We can deal with the test suite and these failures after.

rgommers commented on scipy/stats/tests/test_hypotests.py:
```diff
@@ -202,21 +202,21 @@ def test_input_validation(self):
         # Test that default method ('auto') chooses intended method
 
-        np.random.seed(1)
+        rng = np.random.RandomState(1)
```
Just a note that I tried changing this to `default_rng`; that requires changing `.rand` to `.random` below, and then there's a test failure in this line:
```python
assert auto.pvalue != exact.pvalue
```
That looks like a larger change, best left alone (unclear from a quick glance if the test is not robust, but probably yet since the values do come out the same) - using `RandomState` is hence the simpler thing for this PR.

rgommers replied:
Same for several other tests. These tests are not robust to variations in seeds, e.g. this:
```python
assert np.abs(res1.pvalue - res2.pvalue) > 1e-2
```
is clearly not going to be true when the asymptotic method happens to work well enough.

rgommers committed:
MAINT: stats: fix CI failures and address a few review comments

rgommers commented on scipy/_lib/_testutils.py:
```diff
@@ -195,12 +197,17 @@ def test_views(self):
-            args_array.append(np.repeat(self.arguments[arg_idx][0],
-                                        2, axis=0)[::2])
-        self.pythranfunc(*args_array)
+    @pytest.fixture
```
This is a problem - we can't do `import pytest` at the top of this file nor directly under `_TestPythranFunc`, because `pytest` isn't a runtime dependency. 

`_TestPythranFunc` is only used to test a single function (`stats.somersd`), so apparently there's a regular thread safety issue there.

rgommers replied:
Okay I found the problem, it's crashing:
```
$ pixi r test-nogil -t scipy.stats.tests.test_hypotests::TestSomersD -- --parallel-threads=40 -v
scipy/stats/tests/test_hypotests.py::TestSomersD::test_strided Fatal Python error: Aborted

Current thread 0x00007058d8ff96c0 (most recent call first):
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_core/fromnumeric.py", line 57 in _wrapfunc
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_core/fromnumeric.py", line 506 in repeat
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/_lib/_testutils.py", line 201 in test_strided
  File "/home/rgommers/code/tmp/pytest-run-parallel/src/pytest_run_parallel/plugin.py", line 59 in closure
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 992 in run
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1041 in _bootstrap_inner
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1012 in _bootstrap

Thread 0x00007058d97fa6c0 (most recent call first):
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/stats/_hypotests.py", line 689 in _somers_d
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/stats/_hypotests.py", line 886 in somersd
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/stats/tests/test_hypotests.py", line 682 in pythranfunc
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/scipy/build-nogil-install/lib/python3.13t/site-packages/scipy/_lib/_testutils.py", line 203 in test_strided
  File "/home/rgommers/code/tmp/pytest-run-parallel/src/pytest_run_parallel/plugin.py", line 59 in closure
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 992 in run
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1041 in _bootstrap_inner
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1012 in _bootstrap

Thread 0x00007058d9ffb6c0 (most recent call first):
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 316 in _acquire_restore
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 368 in wait
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 394 in wait_for
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 764 in _wait
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 729 in wait
  File "/home/rgommers/code/tmp/pytest-run-parallel/src/pytest_run_parallel/plugin.py", line 57 in closure
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 992 in run
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1041 in _bootstrap_inner
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/threading.py", line 1012 in _bootstrap

Thread 0x00007058da7fc6c0 (most recent call first):
  File "/home/rgommers/code/pixi-dev-scipystack/scipy/.pixi/envs/free-threading/lib/python3.13t/site-packages/numpy/_core/fromnumeric.py", line 52 in _wrapfunc
  File ???Fatal Python error: Segmentation fault
```

A lock in the tests isn't a fix though. This looks like a thread safety issue in Pythran for strided input. What should be done with this is turn the test into a self-contained reproducer and report it as a Pythran issue.

andyfaff commented:
Is there a guide somewhere on what devs need to do to make sure code is threadsafe?

rgommers commented:
> Is there a guide somewhere on what devs need to do to make sure code is threadsafe?

Not yet, it will come after things are stabilizing - and also we'll have a CI job to smoke out issues. SciPy is the first project where we're really using `pytest-run-parallel` at scale, so new issues tend to come up. 

The most common issue of course is using `np.random.seed` & co., but we all understand that one I think. The second most common ones I think are (a) caches, and (b) storing mutable state without making it thread-local. For the latter, that always had to be done with thread-local variables in most cases, but we neglected to do so in a bunch of places.

rgommers commented:
> On this feature branch, I didn't see any signs of hanging/slow performance, but I still observed some test failures (157 in total).

Thanks for testing Tyler. I am seeing those failures as well. I thought at first that this PR was fairly incompletely, but the actual problem is that it relies on https://github.com/Quansight-Labs/pytest-run-parallel/pull/19, which isn't part of the latest release (0.2.0). With that feature include, I'm only seeing a single failure locally.

rgommers commented:
@andyfaff to add to the answer above: there is a detailed guide about porting to free-threading: https://py-free-threading.github.io/porting/. That's certainly worth reading, however in this PR and quite a few others, the problems we're finding are different from the ones covered there and either more basic, specific to parallel testing, or issues that are also present for default (rather than free-threading) builds of Python. Those things aren't covered in a guide yet.

rgommers committed:
BUG: revert `import pytest` usage in `_TestPythranFunc`

rgommers committed:
TST: seed the use of `bootstrap` for determining confidence interval

rgommers commented:
This failure is showing up in the Accelerate CI job:
```
 FAILED scipy/stats/tests/test_continuous_basic.py::test_cont_basic[500-johnsonsu-arg54] - assert np.float64(0.762988202846999) <= np.float64(0.7606411239708051)
```
It's there with or without seeding the call to `stats.bootstrap`, and I can reproduce it locally both with a regular Python 3.12 as well as with free-threading.

rgommers committed:
TST: tweak seed and comment on non-robust stats test

rgommers commented:
The CI job that tests with the minimum NumPy version is failing with:
```
TypeError: SeedSequence expects int or sequence of ints for entropy not RandomState(MT19937)
```
That will probably go away once we bump the minimum version to 1.25.x, so let's do that first (xref gh-22012).

tylerjereddy commented:
I removed the backport label given the cross-listed PR specifically intended for that purpose.

rgommers commented:
This will probably sit here for a bit. The thread-safety fixes were split off into gh-22175. Summary of what's left:
- Waiting for gh-22012 to drop `numpy<1.25` to sort out the `SeedSequence` issue
- Pythran striding segfault needs to be reported upstream and fixed there
- Waiting for `pytest-run-parallel` 0.3.0
- Then rebase this PR, and merge it when CI is green.

rgommers commented:
NumPy 2.2.3 was released a few days ago - all blockers here are resolved I believe. Let's try again to get this in @andfoy?

andfoy commented:
@rgommers, it seems that the test failures are still present

andfoy committed:
TST: mark matplotlib-dependent tests as thread_unsafe

rgommers committed:
TST: stats: change to `default_rng` to prevent issues with `random_state` wrapper

rgommers commented:
I looked into the `SeedSequence` failures in more detail. I don't remember if I had a reason to suggestion the numpy 1.23 -> 1.25 minimum version bump would resolve them. The actual problem which gives these errors:
```
TypeError: SeedSequence expects int or sequence of ints for entropy not RandomState(MT19937)
```
is that the machinery for deprecating `random_state` keyword (`_transition_to_rng` in `_lib/_util.py`) tries to do some juggling and ends up calling:
```python
np.random.default_rng(rng)
```
That works if `rng` is a `numpy.random.Generator`, but fails with numpy <2.2.0 if it's a `numpy.random.RandomState` instance. Luckily that means the fix is simple: change to `Generator`. 

I'll push a fix now to re-run CI. There is still an issue with that test, already marked with a TODO in this PR. I'll have a closer look at that.

rgommers committed:
TST: stats: found a seed that passes the unreliable `test_continuous_basic` checks

If you change the seed, different distributions fail. The distributions are
fine, the test is conceptually wrong.

ev-br commented:
e3aa2a8dd87ca3715bd6f62efa46ec7c94a7a6f2 looks good. Might want to add a description of the new parameter, or might even keep as is. It not that the parameter is going to be of a very frequent use anyway.

j-bowhay commented on scipy/stats/_kde.py:
```diff
@@ -364,7 +364,7 @@ def integrate_box_1d(self, low, high):
         return value
 
-    def integrate_box(self, low_bounds, high_bounds, maxpts=None):
+    def integrate_box(self, low_bounds, high_bounds, maxpts=None, rng=None):
```
If were adding a new argument maybe we should consider making in kwarg only?

rgommers replied:
Good point, included in gh-22624.

rgommers committed:
TST: stats: fix failures for `bootstrap_against`

rgommers commented:
Okay, this is ready now. The one CI failure is a timeout.

This PR is very likely to accumulate new merge conflicts, so I'd prefer to merge it tomorrow or so. It's test-only and pretty safe to merge, despite its large size. Does anyone want/need more time for this one? Ping @mdhaber since this is most likely to overlap with some of your PRs (touching the same tests perhaps).

andfoy commented:
@rgommers it seems that something went wrong during the rebase process! I would appreciate if you still have the local history, sorry for messing up!

rgommers commented:
Okay, fixed and rebased to get rid of the merge commit. CI is happy again.

rgommers reviewed (approved):
Okay, all seems happy, in it goes. Thanks @andfoy & reviewers!
